### PR TITLE
Execution API spec

### LINKED ISSUE #445: Design the Execution Management Functionality
### Summary
In addition to the standard language server functionality that we aim to provide over time (compile-time diagnostics, auto-completion, and so on), the nature of Enso and its visual environment means that we need to 

### Value
This functionality is the main building block for letting the GUI get updates on watched portions of the program.

### Specification

- [ ] Work carefully to design the exact messages needed to support the following functionality. These messages may be able to be performed on top of `executeCommand`, or may require custom protocol extensions. This is not yet known:
  + [ ] Execution of a function with provided arguments.
  + [ ] Execution of a function from a given call site (stack position and code position).
  + [ ] Attach an execution listener to an arbitrary code span.
  + [ ] Detach an execution listener by ID.
  + [ ] Implement heartbeat messages for execution listeners. If a heartbeat response isn't received before some time-out, the language server should detach the listener. 
  + [ ] Force cache invalidation for arbitrary code spans.
  + [ ] Attach an automatic execution request.
  + [ ] Detach an automatic execution request.
  + [ ] Redirect `stdout`/`stdin`/`stderr` to and from the IDE, and exactly how this should work.
- [ ] For each of the above messages, ensure that they are specified in the design document. 

### Acceptance Criteria & Test Cases

- We have a coherent design for how this portion of the integration is going to work.

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Specifies the first messages in the execution API
Closes #445 
The issue of redirecting IO is ignored for now, should be handled with the visualizations API.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: doc/design/engine/engine-services.md [hunk 1]
```diff
@@ -1065,20 +1076,35 @@ TBC
 ### Capabilities
 The capability management features work with the following capabilities.
 
-#### `capability/canEdit`
+#### `text/canEdit`
 This capability states that the capability has the ability to perform both
 `text/applyEdit` and `text/save` for the specified file.
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]: I think it's worth creating a section per capability that states which messages are enabled and/or disabled by the capability. See below. 

```suggestion
This capability states that the client has the ability to make edits to a given
file.

##### Enables

- `text/applyEdit`
- `text/save`

##### Disables
None
```

Please do this (or something similar) for the remaining capabilities.

File: doc/design/engine/engine-services.md [hunk 2]
```diff
@@ -1696,6 +1722,209 @@ null
 ##### Errors
 TBC
 
+### Execution Management
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]: This should have a brief blurb about what execution management means.
```suggestion
### Execution Management
The execution management portion of the language server API deals with exposing
fine-grained control over program and expression execution to the clients of
the language server. This is incredibly important for enabling the high levels
of interactivity required by Enso Studio.
```

File: doc/design/engine/engine-services.md [hunk 3]
```diff
@@ -1696,6 +1722,209 @@ null
...
+
+#### Types
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]:
```suggestion
#### Types
The execution management API exposes a set of common types used by many of its 
messages.
```

File: doc/design/engine/engine-services.md [hunk 4]
```diff
@@ -1696,6 +1722,209 @@ null
...
+An identifier used for execution contexts.
+
+```typescript
+type ContextId = UUID;
+```
+
+##### `StackItem`
+A representation of an executable position in code, used by the execution APIs.
+
+`MethodCall` is a call performed at the top of the stack, to initialize the
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]: I don't see `MethodCall` in the definition at all. Do you mean `ExplicitCall`?

File: doc/design/engine/engine-services.md [hunk 5]
```diff
@@ -1696,6 +1722,209 @@ null
...
+}
+```
+
+##### Result
+```typescript
+null
+```
+
+##### Errors
+- [`AccessDeniedError`](#accessdeniederror) when the user does not hold the `executionContext/canModify` capability for this context.
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]: Please wrap this line.

File: doc/design/engine/engine-services.md [hunk 6]
```diff
@@ -1696,6 +1722,209 @@ null
...
+```
+
+##### Result
+```typescript
+null
+```
+
+##### Errors
+- [`StackItemNotFoundError`](#stackitemnotfounderror) when the request stack item could not be found.
+- [`AccessDeniedError`](#accessdeniederror) when the user does not hold the `executionContext/canModify` capability for this context.
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]: Line wrapping.

File: doc/design/engine/engine-services.md [hunk 7]
```diff
@@ -1696,6 +1722,209 @@ null
...
+}
+```
+
+##### Result
+```typescript
+null
+```
+
+##### Errors
+- [`AccessDeniedError`](#accessdeniederror) when the user does not hold the `executionContext/canModify` capability for this context.
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md]: Wrapping

File: doc/design/engine/engine-services.md [hunk 8]
```diff
@@ -1706,7 +1921,7 @@ This error signals generic file system errors.
 ```typescript
 "error" : {
   "code" : 1000,
-  "message" : "File '/foo/bar' exists but is a directory"
+  "message" : String
```
[the remark this answers is not in the document]
Comment (kustosz) [reply doc/design/engine/engine-services.md]: yes

**Pull request merged**